### ISSUE TITLE
Atomic write detection breaks event ordering.

### ISSUE THREAD

**Issue by DominoPivot** [paulmillr/chokidar#1370]
When atomic write detection is enabled, as is the default, `"unlink"` events are slightly delayed. If an `"add"` event would occur in that time, a `"change"` event is emitted instead immediately, and the pending `"unlink"` event is discarded. This small delay causes the order of events to become incoherent.

**Versions**
Chokidar 4.0.1
Node.js v22.4.1
Windows 10.0.19045.4894

**Example**

I didn't produce full examples because #1361 compounds with this, making it harder to demonstrate. And because I want to file this issue before I go to bed 😅. Hopefully this is clear enough.

Suppose we have folder A which contains files 1, 2, 3 and so on. We then run the following:

    await rm("A", { recursive: true });
    writeFile("B", "...");

In this situation, despite being guaranteed to happen last, the creation of file B would be detected while deletions are still pending, so chokidar would emit:

    ("add", "B")
    ("unlink", "A/1")
    ("unlink", "A/2")
    ("unlink", "A/3")
    ...

That is to say, chokidar's output contradicts two logical assumptions about the code:
- all deletions occur before any write
- all operations on A occur before operations on B

Preserving order would likely require keeping a buffer of all events that are detected, and pausing all emissions whenever an `"unlink"` is encountered too early to determine if it is part of an atomic write or not. Definitely not a trivial change.

**Comment by paulmillr** [comment]
that’s the whole point of atomic.

rm a.js
mv a.js.swp a.js

**Comment by DominoPivot** [comment]
😕? I'm not sure if you read or understood what I wrote. I know what atomic writes are, and their detection is working.

But the way atomic change detection is implemented in chokidar messes with the order in which events are emitted, even those unrelated to atomic writes. If I delete an output directory before triggering a rebuild of a project, I certainly don't expect the file watcher to emit the deletions AFTER the addition of new files.

For what is worth, using the `watcher` npm package in place of chokidar fixed all those issues for me.

**Comment by paulmillr** [comment]
If the behavior is the same as in v3, I wouldn’t change it for now. Too complicated. PRs welcome.

**Comment by 43081j** [comment]
It's also possible the file system just emits them this way

If you ever get chance, please do make a repro in V3 and V4

**Comment by DominoPivot** [comment]
It seems `fsPromises.rm` can actually resolve before unlink events are fired, at least when running recursively. 🙃
Using `fsPromises.readdir` and `fsPromises.unlink` manually instead seems to have solved a bunch of issues I had.

TL;DR chokidar's default behavior is reasonable and the issue was elsewhere, sorry for the noise.
